chain/chain/src/chain.rs

```diff
@@ -3308,28 +3438,15 @@ impl<'a> ChainUpdate<'a> {
             }
         }
 
-        let mut apply_chunk_work = Vec::new();
-
-        // Always apply state transition for shards in the current epoch
-        apply_chunk_work.extend(self.apply_chunks_preprocessing(
-            me,
-            block,
-            &prev_block,
-            ApplyChunksMode::ThisEpoch,
-        )?);
-
-        // If we have the state for the next epoch already downloaded, apply the state transition for the next epoch as well,
-        //    otherwise put the block into the permanent storage to have the state transition applied later
-        if is_caught_up {
-            apply_chunk_work.extend(self.apply_chunks_preprocessing(
-                me,
-                block,
-                &prev_block,
-                ApplyChunksMode::NextEpoch,
-            )?);
+        // If we have the state for shards in the next epoch already downloaded, apply the state transition
+        // for these states as well
+        // otherwise put the block into the permanent storage, waiting for be caught up
+        let apply_chunk_work = if is_caught_up {
+            self.apply_chunks_preprocessing(me, block, &prev_block, ApplyChunksMode::IsCaughtUp)?
         } else {
             self.chain_store_update.add_block_to_catchup(prev_hash, *block.hash());
-        }
+            self.apply_chunks_preprocessing(me, block, &prev_block, ApplyChunksMode::NotCaughtUp)?
+        };
```

Comment:
Alternative way to write this:

```rust
        let mode = if is_caught_up {
            ApplyChunksMode::IsCaughtUp
        } else {
            self.chain_store_update.add_block_to_catchup(prev_hash, *block.hash());
            ApplyChunksMode::NotCaughtUp
        };
        let apply_chunk_work = self.apply_chunks_preprocessing(me, block, &prev_block, mode)?;
```

Not sure if it's actually better -- I personally default to deduping such things just as a rule of thumb.